tests/Stub/TestLogger.php

```diff
@@ -0,0 +1,194 @@
+ * @method bool hasErrorRecords()
+ * @method bool hasWarningRecords()
+ * @method bool hasNoticeRecords()
+ * @method bool hasInfoRecords()
+ * @method bool hasDebugRecords()
+ *
+ * @method bool hasEmergencyThatContains($message)
+ * @method bool hasAlertThatContains($message)
+ * @method bool hasCriticalThatContains($message)
+ * @method bool hasErrorThatContains($message)
+ * @method bool hasWarningThatContains($message)
+ * @method bool hasNoticeThatContains($message)
+ * @method bool hasInfoThatContains($message)
+ * @method bool hasDebugThatContains($message)
+ *
+ * @method bool hasEmergencyThatMatches($message)
+ * @method bool hasAlertThatMatches($message)
+ * @method bool hasCriticalThatMatches($message)
+ * @method bool hasErrorThatMatches($message)
+ * @method bool hasWarningThatMatches($message)
+ * @method bool hasNoticeThatMatches($message)
+ * @method bool hasInfoThatMatches($message)
+ * @method bool hasDebugThatMatches($message)
+ *
+ * @method bool hasEmergencyThatPasses($message)
+ * @method bool hasAlertThatPasses($message)
+ * @method bool hasCriticalThatPasses($message)
+ * @method bool hasErrorThatPasses($message)
+ * @method bool hasWarningThatPasses($message)
+ * @method bool hasNoticeThatPasses($message)
+ * @method bool hasInfoThatPasses($message)
+ * @method bool hasDebugThatPasses($message)
+ */
+class TestLogger extends AbstractLogger
+{
+    /**
+     * @var array
+     */
+    public $records = array();
+
+    public $recordsByLevel = array();
+
+    /**
+     * @inheritdoc
+     */
+    public function log($level, $message, array $context = array())
+    {
+        if (!in_array($level, $this->getLogLevels(), true)) {
+            throw new InvalidArgumentException(sprintf('Log level "%1$s" is not valid', $level));
+        }
+
+        $record = array(
+            'level' => $level,
+            'message' => $this->formatMessage($message, $level, $context),
+            'context' => $context,
+        );
+
+        $this->recordsByLevel[$record['level']][] = $record;
+        $this->records[] = $record;
+    }
+
+    public function interpolateContext($message, array $context)
+    {
+        return preg_replace_callback('!\{([^\}\s]*)\}!', function ($matches) use ($context) {
+            $key = isset($matches[1]) ? $matches[1] : null;
+            if (array_key_exists($key, $context)) {
+                return $context[$key];
+            }
+
+            return $matches[0];
+        }, $message);
+    }
+
+    public function formatMessage($message, $level, array $context)
+    {
+        $message = $this->interpolateContext($message, $context);
+        $message = "$level $message";
+
+        return $message;
+    }
+
+    public function getLogLevels()
+    {
+        $reflection = new ReflectionClass('Psr\Log\LogLevel');
+        $constants = $reflection->getConstants();
+
+        return $constants;
+    }
+
+    public function getRecords()
+    {
+        return $this->records;
+    }
```

Comment:
All this methods are new and should be documented as new features.

Reply:
If I were to do it, I would probably add that information to the changelog. Is this what you mean?

On the other hand, like I suggested in another comment, this looks like a totally internal class (I think you mentioned yourself that it is not part of the API that can break etc), and so perhaps it doesn't matter.